lib/ukdebug: Add gdb stub to unikraft

thass0 opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

The GDB stub is invoked by debug traps. It allows an instance of GDB (the GDB host) on another system outside the kernel to connect to and debug the running kernel. The GDB stub and the host communicate over a serial connection using the GDB remote protocol. Code implementing the protocol was written some time ago by @marcrittinghaus. This pull request ports Marc's code to the current version of Unikraft and continues on his work.

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): x86_64, ARM64
 - Platform(s): `kvm`
 - Application(s): N/A


### Additional configuration

The GDB stub can be enabled in the `kconfig` menu of `ukdebug`. There are two new kernel parameters that control the GDB stub. They are described in the commit message of this commit: f66258272fd18bdebc04889f27ad534aeed5ccbf.

### Description of changes

The stub is registered as an event handler for the `UKARCH_TRAP_DEBUG` event. So the stub is entered if, e.g., a breakpoint is hit. The event handler is implemented in `lib/ukdebug/arch/$(ARCH)/gdbsup.c`.  Inside the handler, the stub enters a loop. It reads incoming packets sent by the gdb host, acts on these packets and responds. See `gdb_main_loop` in `lib/ukdebug/gdbstub.c`. Some packets instruction the gdb stub to resume execution. The stub leaves the loop if such a packet is received, or on error. Most of `gdbstub.c` is code parsing the incoming packets. The key packets from the gdb remote protocol that the stub supports are: `M` (write memory at address), `m` (read memory at address), `G` (write general registers), `g` (read general registers), `s` (single step), and `c` (continue execution). The underlying implementations are found in the `gdbsup` files.

GitHub-Depends-On: #1464
GitHub-Depends-On: #1481

thass0 committed:
lib/ukdebug: Add breakpoints to set up and connect to stub

If the gdb stub is enabled, it registers a function in the inittab
that tries to connect to the gdb host. This function first checks
if the console with the ID from `debug.gdbcon` is available. If so,
this function executes a breakpoint instruction to enter the gdb
stub. If the console is not available, the breakpoint is skipped.

Checkpatch-Ignore: LONG_LINE_STRING
Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
lib/ukdebug: Disable gdb's ack mode if the stub is connected

Using acknowledgment packets only makes sense on unreliable connections.
But if the connection is reliable, disabling acknowledgment packets
makes the communication between gdb and the stub much faster.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
lib/ukdebug: Communicate with gdb host through `ukconsole`

The gdb stub now uses the `ukconsole` interface to talk
to gdb. That's why the stub also depends on `LIBUKCONSOLE`.
Without `ukconsole` providing a serial connection, the
stub is entirely useless.

The gdb stub communicates with the gdb host using one of
the console devices that are registered with `ukconsole`.
The ID of the console device that the stub should use is
specified in the `debug.gdb_cons` parameter. The stub will
clear the `UK_CONSOLE_FLAG_STD[OUT|IN]` flags on that device
and keep a reference to it to communicate with the gdb host.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
drivers/uktty: Remove newline conversion from drivers

Turning a `\n` character into a `\r\n` character sequence
breaks the communication between the gdb stub and the gdb
host. The stub will calculate a checksum on the message
without the extra carriage returns but the host will receive
a message with the extra carriage returns that the drivers
add and thus the host will calculate a different checksum
than the stub.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
lib/*: Turn `\n` into `\r\n`

See the TODOs in the diff

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

michpappas committed:
arch: Add SP and PC getters and setters for struct __regs

Add architecture-specific getters/setters for the stack pointer
and program counter of a `struct __regs`. ELR_EL1 is used as the
PC on Arm since `struct __regs` is used in exception context.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>
Signed-off-by: Michalis Pappas <michalis@unikraft.io>
Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>
Co-authored-by: Thassilo Schulze <thassilo@unikraft.io>

thass0 committed:
lib/ukdebug, plat/common: Invalidate i-cache if GDB writes memory

On ARM we need to always invalidate the i-cache if the GDB stub
writes somewhere into memory because the stub might change the
executable code.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

marcrittinghaus committed:
lib/ukdebug: Introduce gdb stub beginning with send/receive functions

This commit starts introducing the gdb stub with the functions that
allow the stub to send data to and receive data from the gdb host.
The actual transmission of data isn't implemented yet. That's why
the `gdb_dbg_putc` and `gdb_dbg_getc` functions are stubs for now.

Checkpatch-Ignore: MACRO_WITH_FLOW_CONTROL
Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

marcrittinghaus committed:
lib/ukdebug, arch/arm: Add main loop to handle incoming packets

The main loop of the gdb stub receives packets, calls the appropriate
handlers and responds to the gdb host. The loop is entered after a
debug trap has been caught.  The code in `lib/ukdebug/arch/*` implements
the trap handling for the ARM64 and x86_64 architectures.

Checkpatch-Ignore: COMPLEX_MACRO
Checkpatch-Ignore: UNNECESSARY_PARENTHESES
Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

thass0 committed:
lib/uknofault, lib/uktest: Fix recursive dependencies

The gdb stub selects `uknofault` to access memory in a safe way.
But making the stub select `uknofault` creates two circular dependencies.
There is a similar issue with `uktest`. This commit works around these
circular dependency issues by switching from `select` to `depends on`.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

marcrittinghaus committed:
lib/ukdebug: Add continue and single step packets to gdb stub

These packets implement the "single step" and "continue" operations
in gdb.

Checkpatch-Ignore: UNNECESSARY_PARENTHESES
Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

marcrittinghaus committed:
lib/ukdebug: Add memory read and write packets to gdb stub

These packets allow the gdb host to read and write memory on the
target system. This capability is used to implement breakpoints for
example.

There's an edge-case that the implementation of `gdb_arch_write_memory`
must handle. Assume virtual memory is enabled with `ukvmem` and consider
the following sequence of events:

1. the system starts and the gdb stub is entered early on in the boot
   sequence
2. the gdb stub writes a breakpoint to some location in the program's
   text
3. execution is resumed. Now virtual memory is initialized. The location
   in the program's text that the stub wrote to is marked read-only
4. the breakpoint that the stub created is hit. The stub has to write
   to the read-only page to remove the breakpoint and resume exeuction

This edge-case is handled by using addresses from `ukplat_page_kmap`
instead of the addresses that are passed to `gdb_arch_write_memory`.

Checkpatch-Ignore: UNNECESSARY_PARENTHESES
Co-authored-by: Thassilo Schulze <thassilo@unikraft.io>
Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>
Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

marcrittinghaus committed:
lib/ukdebug: Add register read and write packets to gdb stub

These packets allow the gdb host to read and write the registers
on the target system.

Source of import (lib/ukdebug/arch/arm64/target.xml):
Repository: https://gitlab.com/qemu-project/qemu
Tag: v9.1.0  (fd1952d)
Path: gdb-xml/aarch64-core.xml

NOTE: The `target.xml` file from the QEMU source tree does not
strictly adhere to the gdb remote procol docs because it doesn't
include a `target` element. But in practice this element seems
to be optional.

Source of import (lib/ukdebug/arch/x86_64/target.xml):
Repository: https://gitlab.com/qemu-project/qemu
Tag: v9.1.0  (fd1952d)
Path: gdb-xml/i386-64bit.xml

NOTE: We aim to support the `org.gnu.gdb.i386.core` feature with
the target description for x86_64. The GDB docs specify the minimal
list of registers that must be included in the description for this
feature. See
https://sourceware.org/gdb/current/onlinedocs/gdb.html/i386-Features.html
The x86_64 target description that we have imported from QEMU includes
more register definitions than required by this feature. At present, we
do not support getting the extra registers. That's why any registers
that are included in the target description from QEMU but not required
by the `org.gnu.gdb.i386.core` feature have been removed. The removed
register definitions are: `ss_base` through `k_gs_base`, `cr0`, `cr2`,
`cr3`, `cr4`, `cr8`, `efer`, `xmm0` through `xmm15`, `mxcsr`.

Checkpatch-Ignore: LONG_LINE
Checkpatch-Ignore: COMMIT_LOG_USE_LINK
Co-authored-by: Thassilo Schulze <thassilo@unikraft.io>
Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>
Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

marcrittinghaus committed:
lib/ukdebug: Add `q` commands to gdb stub for target information

The two `q` commands (`qSupported` and `qXfer:features:read:target.xml`)
allow the gdb host to retrieve information about the target system
that's being debugged.

Checkpatch-Ignore: AVOID_EXTERNS
Checkpatch-Ignore: UNNECESSARY_PARENTHESES
Checkpatch-Ignore: SPDX_LICENSE_TAG
Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

marcrittinghaus committed:
lib/ukdebug: Add `vCont` packet to gdb stub

This packet is used by the gdb host as a replacement for the
`c`, `s`, `C` and `S` packets. `vCont` has some more features
than these packets on their own. These features are related to
multi-threading support for the most part. We don't support
the multi-threading features of the gdb remote protocol, so
the `vCont` packet isn't a real improvement over the `c`,
`s`, `C` and `S` packets. But the gdb host still prefers to
use it.

Co-authored-by: Thassilo Schulze <thassilo@unikraft.io>
Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>
Signed-off-by: Marc Rittinghaus <marc.rittinghaus@unikraft.io>

michpappas commented on lib/Config.uk:
```diff
@@ -29,6 +29,10 @@ config HAVE_SYSCALL
 
+config HAVE_DEBUGGER
+       bool
+       default n
```
`default n` not needed

michpappas commented on lib/ukdebug/Config.uk:
```diff
@@ -120,4 +120,14 @@ config LIBUKDEBUG_ALL_TRACEPOINTS
 endif
+
+config LIBUKDEBUG_GDBSTUB
+	bool "Enable GDB debugger stub"
```
Someone pendantic would say that saying "GDB debugger" is like saying "PIN number" or "ATM machine" :) That is autological, as the word "debugger" is already part of the "GDB" acronym, so the term "GDB debugger" expands to "GNU debugger debugger". I think that if you just say "Enable the GDB stub" no one will be confused. Same in the `help` message, you can simply say "connecting GDB to Unikraft".

thass0 replied:
You are right, it is like saying "LCD display"

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -0,0 +1,291 @@
+/*
+ * Author(s): Marc Rittinghaus <marc.rittinghaus@kit.edu>
+ *
+ * Copyright (c) 2021, Karlsruhe Institute of Technology. All rights reserved.
```
I think you can use the short format and retain the KIT copyright. If you made any changes you can also add the usual "Unikraft GmbH and contributors".

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -0,0 +1,291 @@
+
+	/* Wait for packet start character */
+	while ((c = gdb_dbg_getc()) != '$') {
+		if (c < 0)
```
Is that an error code? Then it needs `unlikely()`.

thass0 replied:
Yes, it's an error code. I'm adding the `unlikely()`. I'll also add `unlikely()` to other error checks.

michpappas replied:
It looks like some you missed some error cases both in `gdbstub.c` and the architecture-specific code. Please revise. Most importantly the `GDB_CHECK` macro.

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -0,0 +1,291 @@
+			/* Null-terminate the data */
+			*p = 0;
+			return n;
+		} else if (c == '$') {
```
This should be an abnormal case, so a few warnings here (one for entering this condition and one for giving up) would be useful.

thass0 replied:
Do you mean warning as in `uk_pr_warn` calls? We can't add print calls to the gdb stub because the gdb communication and the print output can go over the same serial connection. If we print something while the gdb host is waiting for a response to a packet, it will interpret whatever we print as the response. This leads to errors.

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -0,0 +1,291 @@
+			GDB_CHECK(gdb_send_ack());
+
+			/* Null-terminate the data */
+			*p = 0;
```
Since it's char based protocol let's use the NUL byte '\0';

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -0,0 +1,291 @@
+
+#define GDB_PACKET_RETRIES 5
+
+#define GDB_CHECK(expr) \
```
Please align backslashes vertically using tabs.

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -0,0 +1,291 @@
+/* '' */
+static __ssz gdb_send_empty_packet(void)
+{
+	return gdb_send_packet(NULL, 0);
```
Does this actually work? For example `gdb_send()` will try to dereference `buf`.

thass0 replied:
Yes, at present it works. `gdb_send()` turns into a call to `uk_console_out_with` and `uk_console_out_with` checks if `len == 0` before dereferencing `buf`. I remember that somewhere in `posix-tty` or `nolibc` there is also code that assumes `len == 0 && buf == NULL` is fine.

thass0 replied:
Update: I wrote the above response looking at the code at the end of the PR. Later, `gdb_send()` turns into a call to `uk_console_out_with`, that's why I wrote the above. You're right that `gdb_send()` tries to dereference `buf` here. But the length if checked for `0` before dereferencing `buf`. This means it shouldn't be a problem

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -0,0 +1,291 @@
+{
+	int r = gdb_dbg_getc();
+
+	return (r < 0) ? r : ((char)r == '+');
```
While reading the calling code I found the logic of the return value unclear. I would expect that this function either returns `bool`, or zero on success.

thass0 replied:
You are right, this is confusing. I changed it to be a function that returns a negative value on error and zero on success. It's easy to introduce stupid bugs if a function returns a boolean on success and a negative value on error. Because IIRC a negative non-zero value is interpreted as `true` in C and only `0` is interpreted as `false`.

michpappas replied:
It would be more clear if you handled the error case separately, i.e.
```c
r = gdb_dbg_getc();
if (unlikely(r < 0))
    return r;

return ((char)r != '+');
```

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -289,3 +320,105 @@ static __ssz gdb_recv_packet(char *buf, __sz len)
+	 * called by the debugger.
+	 */
+#ifdef UKPLAT_LCPU_MULTICORE
+#warning The GDB debugger stub does not support multicore systems
```
"The GDB stub does"

michpappas commented on lib/ukdebug/arch/arm64/gdbsup.c:
```diff
@@ -0,0 +1,132 @@
+ *
+ * Copyright (c) 2021, Karlsruhe Institute of Technology. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
```
Short format, retain copyright, no author.

michpappas commented on lib/ukdebug/arch/x86_64/gdbsup.c:
```diff
@@ -0,0 +1,70 @@
+ *
+ * Copyright (c) 2021, Karlsruhe Institute of Technology. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
```
Short format, retain copyright, no author.

michpappas commented on lib/ukdebug/gdbstub.h:
```diff
@@ -0,0 +1,45 @@
+ *
+ * Copyright (c) 2021, Karlsruhe Institute of Technology. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
```
Short format, retain copyright, no author.

michpappas commented on lib/ukdebug/gdbstub.h:
```diff
@@ -0,0 +1,45 @@
+#include <uk/arch/lcpu.h>
+#include <uk/arch/types.h>
+
+int gdb_dbg_trap(int errnr, struct __regs *regs);
```
I would expect that the return values are declared above this function.

thass0 replied:
I mean, they are declared tow lines below. Doesn't make much of a difference but fine, I switched the order. It's good to be a a little pedantic about style. Otherwise, things end up being kind of messy because lot's of small deviations accumulate. :)

michpappas commented on lib/ukdebug/gdbstub.h:
```diff
@@ -0,0 +1,45 @@
+#define GDB_DBG_CONT    0x0001	/* Continue execution */
+#define GDB_DBG_STEP    0x0002	/* Single-step one instruction */
+
+#endif /* __UKDEBUG_INTERNAL_GDBSTUB_H__ */
```
These `_INTERNAL_` I think are no longer part of our conventions. I also removed them when I was moving the printing facilities from `libukdebug` to `libukprint`. I would rename these to `__UKDEBUG_GDBSTUB_H__`.

michpappas commented on lib/ukdebug/gdbstub.h:
```diff
@@ -0,0 +1,45 @@
+#define __UKDEBUG_INTERNAL_GDBSTUB_H__
+
+#include <uk/arch/lcpu.h>
+#include <uk/arch/types.h>
```
I think `types.h` is not needed.

michpappas commented on lib/ukdebug/arch/arm64/gdbsup.c:
```diff
@@ -0,0 +1,132 @@
+#include <uk/bitops.h>
+#include <uk/compiler.h>
+
+#define MDSCR_EL1_SS		UK_BIT(0)
```
Register definitions must live under `arch/arm/arm64/include/uk/asm/arch.h`

thass0 replied:
Where in the organizational structure of the `arch.h` file should I put these definitions. I don't know the ARM architecture at all. E.g., should the `SPSR` definitions go in the `pstate` section?

thass0 replied:
The `ESR_EL1_EC_BRK64` definition is duplicated

michpappas replied:
>  I don't know the ARM architecture at all. E.g., should the SPSR definitions go in the pstate section?

Regardless of where its values end up upon exception return, `SPSR_EL1` is a system register, so you can add the above definitions in the System Register Definitions section of `arch.h`, at: https://github.com/unikraft/unikraft/blob/staging/arch/arm/arm64/include/uk/asm/arch.h#L307. Similarly you can add `MDSCR_EL1` at: https://github.com/unikraft/unikraft/blob/staging/arch/arm/arm64/include/uk/asm/arch.h#L260

michpappas commented on lib/ukdebug/arch/arm64/gdbsup.c:
```diff
@@ -0,0 +1,132 @@
+	return r;
+}
+
+UK_EVENT_HANDLER(UKARCH_TRAP_DEBUG, gdb_arch_debug_handler);
```
Besides registering a trap handler we should make sure that breakpoints are correctly configured for the given EL and Security State. Specifically the fields of `DBGBCR<n>_EL1` are UNDEFINED in cold reset. See D2.8.4 of DDI0487K.

michpappas replied:
Regarding this one you can add an `gdb_arch_init()` function when you register with inittab.

michpappas commented on lib/ukdebug/arch/arm64/gdbsup.c:
```diff
@@ -0,0 +1,132 @@
+	mdscr |= MDSCR_EL1_SS;
+	mdscr |= MDSCR_EL1_KDE;
+
+	SYSREG_WRITE(mdscr_el1, mdscr);
```
Normally you would need an `ISB` here, but I think it's okay as you will be doing `eret` later.

thass0 replied:
So can I leave it like this?

michpappas replied:
I think yes 👍🏼

michpappas commented on lib/ukdebug/arch/arm64/gdbsup.c:
```diff
@@ -0,0 +1,132 @@
+{
+	__u32 opcode;
+
+	if (!uk_nofault_memcpy((void *)&opcode, (void *)pc, sizeof(opcode),
```
Why does this return "not brk" if memcpy returns zero bytes?

thass0 replied:
I agree, `uk_nofault_memcpy` only succeeds if its return value is equal to the number passed on the `len` parameter. So if `uk_nofault_memcpy(...) != sizeof(opcode)`, it's an error.

michpappas replied:
In the latest update the error is not handled correctly, as the function returns one of the two boolean states. Also you should use `unlikely` here too.

michpappas commented on lib/ukdebug/arch/arm64/gdbsup.c:
```diff
@@ -0,0 +1,132 @@
+			       UK_NOFAULTF_NOPAGING))
+		return 0;
+
+	return ((opcode & BRK_OPCODE) == BRK_OPCODE);
```
Similarly as before, given its name I would expect this function to return `bool`.

thass0 replied:
I didn't know we had a bool type

michpappas replied:
As `libukdebug` depends on `HAVE_LIBC` I guess it's okay to include `stdbool.h`. See also the `Return Values`  section in https://unikraft.org/docs/contributing/coding-conventions, although this may need amending given that, as you said, there is no kernel `__bool` type. @skuenzer I propose that we introduce a `__bool` type.

michpappas commented on lib/isrlib/Config.uk:
```diff
@@ -1,4 +1,4 @@
     bool "isrlib: Interrupt-safe standard routines"
     default n
-    select LIBNOLIBC if !HAVE_LIBC
+    depends on (HAVE_LIBC || LIBNOLIBC)
```
I think we have to look at these dependencies a bit closer, as the root of the problem may be on the original dependencies. As a matter of fact, does `isrlib` really need to select `libnolibc`?

thass0 replied:
`isrlib` uses `libc` headers so I guess the answer is yes. At least `isrlib` needs a `libc` to be present.

michpappas replied:
That's right, but my point is whether that should be the case. In `isrlib` only `SIZE_MAX` and `ULONG_MAX` depend on libc. `stdint.h` can be removed by switching to kernel types. Ultimately the question is whether `isrlib` should depend on libc at all. CC: @skuenzer.

skuenzer replied:
I agree that we can resolve the dependency. A Unikraft version is provided for those. The easiest way to fix is to make sure that these types never appear in any isrlib provided header and internally (in the .c files) the equivalent Unikraft constants are used.

michpappas commented on lib/ukdebug/arch/arm64/target.xml:
```diff
@@ -0,0 +1,50 @@
+<?xml version="1.0"?>
```
Please add traceability information on the origin of the file. E.g. (from 75ec4931e):
```
    Source of import:
    Repository: https://git.musl-libc.org/cgit/musl
    Tag: v1.2.4 (f5f55d65)
    Path: musl/include/sys/random.h
```

thass0 replied:
@marcrittinghaus Do you recall where the `target.xml` file is from?

michpappas replied:
You can do a Google search using a substring from that file.

thass0 replied:
Looks like both the XML files are from QEMU. Updating the commit

thass0 replied:
I removed the original files from Marc's work because we couldn't find the origin of the `target.xml` files that Marc used. I replaced them with the right target descriptions from the QEMU source tree. The commit that  introduces the `target.xml` files contains both the traceability information that you requested and notes on any changes that I had to make.

michpappas commented on lib/ukdebug/arch/arm64/gdbsup.h:
```diff
@@ -0,0 +1,44 @@
+ *
+ * Copyright (c) 2021, Karlsruhe Institute of Technology. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
```
Short form, no author, retain copyright.

michpappas commented on lib/ukdebug/arch/x86_64/gdbsup.h:
```diff
@@ -0,0 +1,44 @@
+ *
+ * Copyright (c) 2021, Karlsruhe Institute of Technology. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
```
Short form, no author, retain copyright.

michpappas commented on lib/ukdebug/arch/arm64/gdbsup.h:
```diff
@@ -0,0 +1,44 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __UKDEBUG_INTERNAL_ARCH_GDBSUP_H__
```
I think `INTERNAL` can go away.

michpappas commented on lib/ukdebug/arch/x86_64/gdbsup.h:
```diff
@@ -0,0 +1,44 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __UKDEBUG_INTERNAL_ARCH_GDBSUP_H__
```
I think `INTERNAL` can go away.

michpappas commented on lib/ukdebug/arch/arm64/gdbsup.h:
```diff
@@ -0,0 +1,44 @@
+#ifndef __UKDEBUG_INTERNAL_ARCH_GDBSUP_H__
+#define __UKDEBUG_INTERNAL_ARCH_GDBSUP_H__
+
+#include <uk/arch/types.h>
```
`<uk/arch/types.h>` needed here?

michpappas commented on lib/ukdebug/arch/x86_64/gdbsup.h:
```diff
@@ -0,0 +1,44 @@
+#ifndef __UKDEBUG_INTERNAL_ARCH_GDBSUP_H__
+#define __UKDEBUG_INTERNAL_ARCH_GDBSUP_H__
+
+#include <uk/arch/types.h>
```
`<uk/arch/types.h>` needed here?

michpappas commented on lib/ukdebug/arch/arm64/gdbsup.h:
```diff
@@ -0,0 +1,44 @@
+#include <uk/arch/types.h>
+#include <uk/arch/lcpu.h>
+
+static inline void gdb_arch_set_ip(unsigned long ip, struct __regs *regs)
```
We also introduce macros to set the stack pointer in the vfork / execve PR: https://github.com/unikraft/unikraft/pull/1386/commits/cc90a90c9476115ac374ec0896a6a6d80e180652

Maybe these should go there too and we merge a separate PR that this can depend on. @mogasergiu thoughts?

michpappas replied:
After an offline talk with @mogasergiu: You can take https://github.com/unikraft/unikraft/pull/1386/commits/cc90a90c9476115ac374ec0896a6a6d80e180652, and add the IP getter there. Make sure to add your `Co-Authored-By`, and remove mine; I only had to include it cause I was including it in my PR.

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -135,6 +136,43 @@ static int gdb_hex2int(char hex)
 	return -EINVAL;
 }
 
+static unsigned long gdb_hex2ulong(const char *buf, __sz buf_len, char **endptr)
```
This should be probably implemented as `strtoul()` in `isrlib`.

thass0 replied:
I'm not sure how important that is.

EDIT: Or if it's worth the effort

michpappas replied:
IMO it is, as these functions are in the scope of the `isrlib`, and we should avoid code duplication in general. The same probably applies to `strntoull()` that is defined in `libukparam`. @skuenzer thoughts?

skuenzer replied:
Hum... a good question. I remember a design choice was to have a self-contained GDB code base as much as possible to avoid to rely on potential broken libc functions and make sure debugging always works, even of those. Ideally, the GDB stub is independent of any libc (also because weird dependencies to TLS sometimes).

thass0 replied:
I would stick to the self-contained code. Both to keep the dependencies of the gdb stub minimal and because of time

skuenzer replied:
Yes, I agree with this.

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -0,0 +1,291 @@
+	return -EINVAL;
+}
+
+static __ssz gdb_hex2mem(char *mem, __sz mem_len, const char *hex, __sz hex_len)
```
This together with `gdb_hex2int()` should be `atoi()`. Since `isrlib` is a dependency, it would make sense to add an `atoi()` implementation there instead.

CC: @skuenzer

thass0 replied:
Same thing as before: https://github.com/unikraft/unikraft/pull/1479#discussion_r1734730968

michpappas replied:
Ok, let's keep the discussion there.

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -0,0 +1,291 @@
+			 */
+			p = buf;
+			n = 0;
+		} else {
```
According to https://sourceware.org/gdb/current/onlinedocs/gdb.html/Overview.html#Overview sequence ids are delimited with `:` and are deprecated since GDB-5, and as such they must not be accepted by the stub. Given that GDB-5 is a very old version I think it's okay to omit a check for `:`.

thass0 replied:
OK so you're saying I should keep the receive loop as is?

michpappas replied:
Yes I think it's okay to leave is as is 👍🏼

michpappas commented on lib/ukdebug/arch/arm64/gdbsup.c:
```diff
@@ -130,3 +138,56 @@ static int gdb_arch_debug_handler(void *data)
 
 UK_EVENT_HANDLER(UKARCH_TRAP_DEBUG, gdb_arch_debug_handler);
+
+__ssz gdb_arch_read_memory(unsigned long addr, __sz len,
```
I don't see any arch-specific code here, and the implementation looks the same as in `x86_64/gdbsup.c`.  If that's correct these should be moved to `gdbstub.c` instead. 

Update: I see that later you add the i-cache invalidation in `arm64`, so perhaps that's the reason behind this duplication. In that case I would suggest that we move this code to `stub.c` and add an arch-specific call for the i-cache that is implemented in x86_64 as a NOP.

thass0 replied:
I agree, will do that

michpappas commented on lib/ukdebug/arch/arm64/gdbsup.c:
```diff
@@ -139,6 +139,72 @@ static int gdb_arch_debug_handler(void *data)
+	{__offsetof(struct __regs, spsr_el1), 8}
+};
+
+#define GDB_REGISTER_MAP_NUM ((int)(sizeof(gdb_register_map) / \
```
`#define GDB_REGISTER_MAP_NUM        ARRAY_SIZE(gdb_register_map)`. Or even better `GDG_REGISTER_MAP_COUNT`.

michpappas commented on lib/ukdebug/arch/arm64/gdbsup.c:
```diff
@@ -139,6 +139,72 @@ static int gdb_arch_debug_handler(void *data)
+__ssz gdb_arch_read_register(int regnr, struct __regs *regs,
+			     void *buf, __sz buf_len __maybe_unused)
+{
+	if (regnr < 30) {
```
`if (regnr < 30) { /* x0-x29 */`

michpappas commented on lib/ukdebug/arch/arm64/gdbsup.c:
```diff
@@ -139,6 +139,72 @@ static int gdb_arch_debug_handler(void *data)
+	if (regnr < GDB_REGISTER_MAP_NUM) {
+		UK_ASSERT(buf_len >= gdb_register_map[regnr].length);
+
+		memcpy_isr(buf,
```
That initially looked strange as all registers defined in the xml are of fixed size, but think the reason we're using `memcpy()` here is forward compatibility for FPU / SIMD. Let it be then.

michpappas commented on lib/ukdebug/arch/arm64/target.xml:
```diff
@@ -0,0 +1,50 @@
+
+<!-- AArch64 -->
+
+<!DOCTYPE target SYSTEM "gdb-target.dtd">
```
Not necessarily in the scope of this PR, but should we be adding the definitions from `aarch64-fpu` as well? https://android.googlesource.com/platform/external/qemu/+/emu-master-dev/gdb-xml/aarch64-fpu.xml

michpappas replied:
I think we don't have time to properly implement this now, let's leave it for later. Resolving.

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -602,6 +671,125 @@ static int gdb_handle_step_sig(char *buf, __sz buf_len,
+	return 0;
+}
+
+/* qXfer:object:read:annex:offset,length */
```
Should probably be `qXfer:features:read`

thass0 replied:
The way the gdb docs describe it is there is `‘qXfer:object:read:annex:offset,length’` and this has different forms where `object` is replaced by some other word like `features` or `btrace` for example. Probably this handler is meant to cover all of those if we add any more forms of this packet. See https://sourceware.org/gdb/current/onlinedocs/gdb.html/General-Query-Packets.html#qXfer-target-description-read. I'd say let's keep it as is.

michpappas replied:
Okay, so this function should later be populated to handle more object types 👍🏼

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -75,14 +80,43 @@ struct gdb_cmd_table_entry {
 {
-	return -ENOTSUP;
+	if (!gdb_backing_console)
+		return -ENOTSUP;
```
`unlikely` and `-ENODEV` maybe?

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -75,14 +80,43 @@ struct gdb_cmd_table_entry {
+static struct uk_console *gdb_backing_console;
+
+#define GDB_BACKING_CONSOLE_NONE	((__u16)(-1))
+static __u16 gdb_backing_console_id = GDB_BACKING_CONSOLE_NONE;
```
While reading that I found that the word "backing" adds confusion, especially in the above comment that mentions "this console" and "backing console" made me think they are different devices. IMO `gdb_console` would be enough, but that's a bit subjective so feel free to ignore if you think otherwise.

thass0 replied:
I agree that the comment is poorly phrased by using the terms "this console" and "backing console" to refer to the same thing. I'll revise it. But I think using longer names for global variables is generally a good thing. Makes them stand out a little more and adds more context.

michpappas replied:
> But I think using longer names for global variables is generally a good thing. Makes them stand out a little more and adds more context.

That's true, globals should be more descriptive that temporary vars. However longer names are good as long they provide useful info. In this case `gdb_backing_console` has pretty low entropy, as there is no other gdb console and also there is no such thing as a non-backing console. Also under that logic you could say "backing memory" in almost all contexts, but no one does that :D

thass0 replied:
>  there is no other gdb console and also there is no such thing as a non-backing console.

There's the `gdb_virt_console`

michpappas replied:
You're right, I totally forgot about that. In that case it makes total sense to differentiate between the two 👍🏼

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -75,14 +80,43 @@ struct gdb_cmd_table_entry {
+
+	if (uk_console_out_with(gdb_backing_console, &b, 1) == 1)
+		return 0;
+	else
```
You don't need that `else` here:
```c
if (uk_console_out_with(gdb_backing_console, &b, 1) == 1)
		return 0;

return -EIO;
```

michpappas replied:
Looking back at this, you can updated it to use `unlikely`:
```c
rc = uk_console_out_with(gdb_backing_console, &b, 1);
if (unlikely(rc != 1))
	return -EIO;

return 0;
```

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -75,14 +80,43 @@ struct gdb_cmd_table_entry {
+	char b;
+
+	if (!gdb_backing_console)
+		return -ENOTSUP;
```
`unilkely` and `-ENODEV`?

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -75,14 +80,43 @@ struct gdb_cmd_table_entry {
+	if (!gdb_backing_console)
+		return -ENOTSUP;
+
+	/* Buzy loop until we read one byte */
```
s/Buzy/Busy/

thass0 replied:
haha, yes

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -954,6 +984,46 @@ int gdb_dbg_trap(int errnr, struct __regs *regs)
 		return GDB_DBG_CONT;
 	}
 
+	if (!gdb_backing_console) {
```
For readability I would move console selection into a separate function.

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -94,6 +94,28 @@ UK_LIBPARAM_PARAM_ALIAS(gdb_cons, &gdb_backing_console_id, __u16,
 
+/* This is a tristate value controlling the behavior of the gdb virtual
+ * console. It's value is interpreted as follows:
+ * - (> 0): always register the virtual console
```
This comment could be improved a bit, as "register the virtual console" on its own does not mean much (with what?).

In the commit message saying "kernel's console output" instead of "kernel output" in the first sentence would 
make things more clear.

thass0 replied:
It has been addressed. On my github it shows up like this:

```
/* This is a tristate value controlling the behavior of the gdb stub's virtual
 * console. The virtual console implements a console device for `ukconsole`.
 * This console device is called "virtual" because it's not a driver for a
 * real console device. The virtual console sends all console output to the
 * gdb host using the 'O' packet. The backing console is used to transmit this
 * packet like any other. With the virtual console, the backing console can be
 * used for kernel console output and for gdb communication at the same time.
 *
 * The value of this tristate is interpreted as follows:
 * - (> 0): always register the virtual console
[...]
```

michpappas commented on arch/arm/arm64/include/uk/asm/lcpu.h:
```diff
@@ -381,4 +381,9 @@ static inline void ukarch_spinwait(void)
 	/* Intelligent busy wait not supported on arm64. */
 }
 
+static inline void ukarch_dbg_break(void)
```
This must go into `libukdebug/arch/`

thass0 replied:
Eventually this function shouldn't even be visible publicly. `ukdebug` should just register one function that runs at the end of the early init tab and one function that runs at the end of the (late) inittab. Both of these functions include a breakpoint and allow gdb to connect to the stub. We already do this in case of the inittab. We can't do it in case of the early init tab because the functions in the early init tab are executed before the lcpu and interrupts are initialized. That's why I manually inserted the early breakpoint in `ukboot` (at the earliest possible point in the boot sequence). If the lcpu and interrupts can be initialized early enough, we can make `ukarch_dbg_break` private to `ukdebug`.

If you want me to remove the definition in `arch/` now, I can create a new public header in `ukdebug` that contains the `ukarch_dbg_break` definition. Or we wait until `ukarch_dbg_break` becomes private and then remove the definition in `arch/`.

michpappas replied:
So in principle you could define a function `uk_debug_gdb_early_init()` and call that instead. It would also allow you to do more things like checking whether a console has been selected, or calling an arch-specific init function.

michpappas commented on arch/x86/x86_64/include/uk/asm/lcpu.h:
```diff
@@ -169,4 +169,10 @@ static inline void ukarch_x86_cpuid(__u32 fn, __u32 subfn,
 			     : "a"(fn), "c" (subfn));
 }
+
+static inline void ukarch_dbg_break(void)
```
This must go into `libukdebug/arch/`

thass0 replied:
Let's have this discussion here: https://github.com/unikraft/unikraft/pull/1479#discussion_r1734788381

michpappas commented on plat/kvm/arm/setup.c:
```diff
@@ -177,6 +177,14 @@ void __no_pauth _ukplat_entry(void)
 	if (unlikely(rc))
 		UK_CRASH("Failed to initialize bootstrapping CPU: %d\n", rc);
 
+#if CONFIG_HAVE_DEBUGGER
```
This needs to be part of earlytab but, as discussed, needs lcpu and intctlr to be initialized first. Leaving this comment here as a reminder to resolve.

thass0 replied:
That's right. This comment is related: https://github.com/unikraft/unikraft/pull/1479#discussion_r1734788381

michpappas replied:
Regarding this as well as https://github.com/unikraft/unikraft/pull/1479#discussion_r1734788381: I have actually doubts on whether we can provide gdb facilities via earlytab, as the interrupt controller (and possibly lcpu) is not expected to be initialized at early init, as that would defeat the purpose of differentiation between early / late init for drivers. Moreover,  IMO the very existence of early / late GDB entry indicates that the current design should be revised. As these are topics at the architecture level I am leaving this open to discuss with @skuenzer.

thass0 replied:
IMO, the ideal outcome would be: (1) one breakpoint on boot, (2) no spill-over of GDB stub-related code into other parts of the system, and (3) the ability for a user to choose any console device they like for the GDB connection.

For context, the following is why we have the early and late breakpoints. The console drivers all register only a single device with `ukconsole` in the earlytab. If there are further devices, they are registered with `ukconsole` later in the inittab. The current setup with early and late breakpoints allows using any console device for the GDB connection.

michpappas replied:
The main problems here are that:
1. There is no `libuklcpu` yet, and we haven't finalized whether `uk_lcpu_init()` should be called before or after early init (right now it's called after early init). That also includes the ability to trap exceptions.
2. We haven't finalized when should the interrupt controller be initialized (right now it's called after early init). As I said in the previous comment, the initial idea was that early init would execute before we initialize the interrupt controller, and for that reason early devices would also have the chance to be probed at a later point via inittab.
3. `mmap` / `kmap` used by the stub requires `ukplat_mem_init()` has been called already. Not having migrated paging into a library aside, once again we initialize memory after the earlytab, and IMO that is unlikely to be change.

In other words, although we now have an early_init, we haven't fully formalized its boundaries. Given that I don't think we can have gdb facilities that early. Besides, the distance between the current "early"and "late" initialization of gdb is pretty small: The former happens right before we jump to libukboot, while the latter happens on inittab.

CC: @skuenzer

skuenzer replied:
As we discussed offline, decision is to skip gdb entry with early_init and rely on inittab only. Another advantage is that `HAVE_DEBUGGER` can then be removed as well.

michpappas commented on drivers/uktty/ns16550/ns16550.c:
```diff
@@ -144,7 +144,7 @@ static void ns16550_reg_write(__u64 base, __u32 reg, __u32 value)
 }
 
-static void _putc(__u64 base, char a)
+static void ns16550_putc(__u64 base, char a)
```
I would rephrase the commit message title as "Remove newline conversion from drivers".

thass0 replied:
OK :)

michpappas commented on lib/devfs/stdout.c:
```diff
@@ -44,9 +44,45 @@
 #ifdef CONFIG_LIBDEVFS_DEV_STDOUT
 #define DEV_STDOUT_NAME "stdout"
 
+/* TODO: Some consoles require both a newline and a carriage return to
```
Although I understand the motivation, I'm not very keen of this replicated code. It would be cleaner if we handled the tty logic in a common place, even if it's a minimal one that for now it only handles the newline conversion.

thass0 replied:
We plan to add a TTY layer, right? IMO the TTY layer is where this replicated code should go. IIRC I discussed this with @skuenzer and replicating the code is what we settled on for now. Of course, we can decide to move it to a common place now given how often this code is replicated. I think when we touched `uk_scanf` in the `ukconsole` PR, we also decided that it should eventually go into the TTY layer. So that's two items for the TTY layer already

skuenzer replied:
Yes, longer term the place needs to be posix-tty and maybe nolibc for compatibility mode when we build without `posix-tty`/`posix-fdtab`. So I thinnk the duplication for now is okay given that we did not finished architecting TTY handling correctly yet.

thass0 committed:
lib/ukdebug: Add `qAttached` packet to gdb stub

This packet is not mandatory but gdb will probe for
it. Now we respond properly to the probing. This
packet must be updated to support the multiprocess
features of the gdb remote protocol.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

michpappas commented on lib/ukdebug/Config.uk:
```diff
@@ -125,6 +125,7 @@ config LIBUKDEBUG_GDBSTUB
 	bool "Enable GDB debugger stub"
 	depends on (ARCH_X86_64 || ARCH_ARM_64)
 	depends on PLAT_KVM
+	depends on LIBUKCONSOLE
```
If it's not possible to use the gdb stub without enabling it via the cmdline, I would also select `LIBUKLIBPARAM` here.

michpappas replied:
Should that be `depends on` or `select`? Normally we prefer forward dependency resolution.

thass0 replied:
`select` breaks kconfig. Same goes for the `depends on LIBUKLIBPARAM` in the line below `depends on LIBUKCONSOLE`.

skuenzer replied:
I guess it is because the typical circle that ukconsole `selects` ukdebug?

thass0 replied:
Yes, it's because of recursive dependency cycles

skuenzer replied:
Hum, I am not sure yet how can properly fix this. I guess with the ukprint library we will revisit. Can you tag this with a `# FIXME: 'select' currently causes a circular dependency` comment?

thass0 replied:
Yes, that's good.

michpappas commented on lib/ukdebug/Config.uk:
```diff
@@ -120,4 +120,33 @@ config LIBUKDEBUG_ALL_TRACEPOINTS
+	select LIBISRLIB
+	select HAVE_DEBUGGER
+	help
+	  The GDB stub allows connecting the GDB debugger to Unikraft via
```
You could document the cmdline options that are required to use the stub. Ideally also you could add a `README.md` with additional instructions on how to connect to gdb on QEMU.

thass0 replied:
Yes, I will write a README. I can't fit everything that people should know about using the gdb stub into this help message alone.

michpappas commented on lib/ukdebug/Config.uk:
```diff
@@ -120,4 +120,33 @@ config LIBUKDEBUG_ALL_TRACEPOINTS
+	bool "Enable 'no acknowledgments' mode"
+	default y
+	help
+	  The GDB host (the instance of GDB that connects to the stub
```
> "(the instance of GDB that connects to the stub from outside)"

I think "host" is a pretty standard terminology. As a rule of thumb you should write for your audience. In this case, if your audience knows how to debug an OS you should expect they are familiar with basic gdb usage.

thass0 replied:
It's fair to be friendly towards newcomers too. Maybe this is the first time they want to use GDB on an OS. I'll remove it from the `help` message and maybe add it to the README that explains how to use the gdb stub.

michpappas reviewed (changes requested):
Hi @thass0, nice work overall. First round of comments below.

thass0 committed:
lib/ukdebug: Add virtual kernel console to gdb stub

The `debug.gdb_kern` parameter is a tristate. It determines
if the gdb stub should send kernel output to the gdb host
over the gdb remote protocol. A negative value means "auto".
In this case, the gdb stub sends kernel output to the gdb
host if the backing console had the `UK_CONSOLE_FLAG_STD[OUT|IN]`
flags set. "auto" is the default. A value of zero means "off";
kernel output is never sent to the gdb host. A positive value
means "on"; kernel output is always sent to the gdb host.

Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

michpappas commented on lib/ukdebug/Config.uk:
```diff
@@ -120,4 +120,35 @@ config LIBUKDEBUG_ALL_TRACEPOINTS
+config LIBUKDEBUG_GDBSTUB
+	bool "Enable GDB stub"
+	depends on (ARCH_X86_64 || ARCH_ARM_64)
+	depends on PLAT_KVM
```
It's better to avoid adding platform dependencies, what exactly depends on KVM?

thass0 replied:
I don't know if there's a platform dependency, actually. This is from the code that @marcrittinghaus gave me to work with.

thass0 replied:
~Oh, there is a platform dependency: the call to `uk_debug_gdb_early_init` in the KVM platform.~ Nvm this is the wrong direction :joy:

michpappas commented on lib/ukdebug/arch/arm64/gdbsup.c:
```diff
@@ -0,0 +1,179 @@
+			       UK_NOFAULTF_NOPAGING) != sizeof(opcode))
+		return 0;
+
+	return ((opcode & BRK_OPCODE) == BRK_OPCODE);
```
`(opcode & BRK_OPCODE) == BRK_OPCODE` can give false positives. Consider the opcode `0xd42fffff` which will also match. What you need is to apply a mask to `opcode` and then compare, i.e.
```
#define BRK_OPCODE_MASK                0xfff0001fUL
```
and then:
```
return ((opcode & BRK_OPCODE_MASK) == BRK_OPCODE);
```

michpappas commented on lib/ukdebug/Config.uk:
```diff
@@ -120,4 +120,35 @@ config LIBUKDEBUG_ALL_TRACEPOINTS
 endif
+
+config LIBUKDEBUG_GDBSTUB
+	bool "Enable GDB stub"
```
"the GDB stub"

michpappas commented on lib/ukdebug/Config.uk:
```diff
@@ -120,4 +120,35 @@ config LIBUKDEBUG_ALL_TRACEPOINTS
+
+if LIBUKDEBUG_GDBSTUB
+config LIBUKDEBUG_GDB_NO_ACK
+	bool "Enable 'no acknowledgments' mode"
```
Maybe `Disable acknowledgements`, or even have acknowledgements disabled by default and provide an `Enable acknowledgements` option?

thass0 replied:
I switched to "Always use acknowledgment packets".

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -0,0 +1,1247 @@
+		return 0;
+	}
+
+	uk_pr_info("Waiting for debugger (late) ...\n");
```
This is reached even if we don't enable gdb via the cmdline. In `x86_64` it will only result into printing the "Waiting..." line and continuing, but in `arm64` execution will stop. You need to check whether a gdb console has been set.

michpappas commented on lib/ukdebug/gdbstub.c:
```diff
@@ -0,0 +1,1247 @@
+
+static int gdb_setup_backing_console(void)
+{
+	/* Ignore breakpoints if the stub is compiled in but not
```
Both this and the next check are breakpoint-specific, so they should be part of the relevant function.

thass0 replied:
I don't think there's a better place for these checks. The variables related to the `backing_console` are all local to the `gdbstub.c` file. The checks are common behavior among both architectures. Both of these points mean that we shouldn't move these checks out of `gdbstub.c`.

thass0 replied:
I changed my mind. First, I added a new function, `uk_debug_gdb_early_init`. This function replaces the calls to `ukplat_dbg_brk` in `plat/kvm`. Additionally, this function sets up the backing console. This change allowed me to delete the `ukplat_dbg_brk` functions in the `arch/` directory. Second, I moved the calls to `gdb_setup_backing_console` into `gdb_late_breakpoint` and `uk_debug_early_init`. There are no more calls to this function from inside the stub. This means that the setup of the backing console now happens before hitting a breakpoint.

michpappas reviewed (changes requested):
Hi @thass0, here's another round.

michpappas commented on lib/ukdebug/arch/arm64/gdbsup.c:
```diff
@@ -0,0 +1,184 @@
+		ctx->regs->spsr_el1 &= ~SPSR_EL1_SS;
+	}
+
+	return r;
```
Since this function is an event handler it must return one of the values mandated by the uk_event API. See `include/uk/event.h`. Same for `x86_64`.

michpappas reviewed (changes requested):
@thass0 I'm leaving one more comment below. Some other minor pending items:
* Remove the dependency of KVM
* See https://github.com/unikraft/unikraft/pull/1479#discussion_r1751587011
* Fix pending checkpatch error
* Fix pending error on the title of the PR (can be reduced to `libukdebug: `
* See if any of the static analysis issues are valid

skuenzer commented on lib/uknofault/Config.uk:
```diff
@@ -21,6 +21,6 @@ config LIBUKNOFAULT_TEST
 	default n
 	select LIBUKTEST
-	select PAGING
+	depends on PAGING
```
Another surprise: What does PAGING has to do with running unit tests?

skuenzer commented on lib/uktest/Config.uk:
```diff
@@ -1,8 +1,8 @@
-	select LIBNOLIBC if !HAVE_LIBC
-	select LIBUKDEBUG
+	depends on (HAVE_LIBC || LIBNOLIBC)
+	depends on LIBUKDEBUG
```
This change is probably okay given that a libc and ukdebug is typically always part of a build. But I see it as a compromise to resolve the circular dependency issue.

skuenzer commented on lib/uknofault/Config.uk:
```diff
@@ -21,6 +21,6 @@ config LIBUKNOFAULT_TEST
 	bool "Enable tests"
 	default n
 	select LIBUKTEST
-	select PAGING
```
Hum... is here still a circular dependency happening if we have fixed the other ones? If so, probably `depends on` is the fix for this one.

thass0 replied:
Yes, the `select PAGING` also leads to a circular dependency:
```
/home/thasso/Work/oss-unikraft/plat/xen/Config.uk:1:error: recursive dependency detected!
/home/thasso/Work/oss-unikraft/plat/xen/Config.uk:1:    symbol PLAT_XEN depends on HAVE_PAGING
/home/thasso/Work/oss-unikraft/plat/Config.uk:86:       symbol HAVE_PAGING default is visible depending on PAGING
/home/thasso/Work/oss-unikraft/plat/Config.uk:62:       symbol PAGING is selected by LIBUKNOFAULT_TEST
/home/thasso/Work/oss-unikraft/lib/uknofault/Config.uk:20:      symbol LIBUKNOFAULT_TEST depends on LIBUKNOFAULT
/home/thasso/Work/oss-unikraft/lib/uknofault/Config.uk:1:       symbol LIBUKNOFAULT is selected by LIBUKDEBUG_GDBSTUB
/home/thasso/Work/oss-unikraft/lib/ukdebug/Config.uk:124:       symbol LIBUKDEBUG_GDBSTUB depends on LIBUKDEBUG
/home/thasso/Work/oss-unikraft/lib/ukdebug/Config.uk:1: symbol LIBUKDEBUG depends on LIBNOLIBC
/home/thasso/Work/oss-unikraft/lib/nolibc/Config.uk:1:  symbol LIBNOLIBC is selected by PLAT_XEN
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
```

skuenzer replied:
okay

skuenzer commented on lib/ukdebug/Config.uk:
```diff
@@ -120,4 +120,12 @@ config LIBUKDEBUG_ALL_TRACEPOINTS
 endif
+
+config LIBUKDEBUG_GDBSTUB
+	bool "Enable the GDB stub"
```
Can we call this option just `"GDB stub"`?

thass0 replied:
Sure :+1:

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -5,22 +5,44 @@
 	do {					\
 		__ssz __r = (expr);		\
-		if (__r < 0)			\
+		if (unlikely(__r < 0))		\
```
A bit weird to introduce the macro in the previous commit without  `unlikely` and add it in the same PR with the following commit. However, just a tiny detail, you can keep it as is.

thass0 replied:
That's a mistake, fixed it

skuenzer commented on lib/ukdebug/Makefile.uk:
```diff
@@ -16,6 +16,8 @@ LIBUKDEBUG_SRCS-y += $(LIBUKDEBUG_BASE)/hexdump.c
 
 LIBUKDEBUG_SRCS-$(CONFIG_LIBUKDEBUG_GDBSTUB) += $(LIBUKDEBUG_BASE)/gdbstub.c|isr
+LIBUKDEBUG_SRCS-$(CONFIG_LIBUKDEBUG_GDBSTUB) += $(LIBUKDEBUG_BASE)/gdbtargetxml.ld
+LIBUKDEBUG_SRCS-$(CONFIG_LIBUKDEBUG_GDBSTUB) += $(LIBUKDEBUG_BASE)/gdbtargetxml.S
```
Can you add the following line:
```
LIBUKDEBUG_GDBTARGETXML_CDEPS += $(LIBUKDEBUG_BASE)/arch/$(CONFIG_UK_ARCH)/target.xml
```
This makes sure that any modifications to the xml file will retrigger a compilation of the assembler file that embeds the xml file.

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -71,7 +71,33 @@ static __u16 gdb_backing_console_id = GDB_BACKING_CONSOLE_NONE;
+ */
+static int gdb_virt_console_mode = -1;
+UK_LIBPARAM_PARAM_ALIAS(gdb_kern, &gdb_virt_console_mode, int,
+			"kernel output on gdb connection");
```
As discussed, we should name this `gdbmode`, or maybe `gdbkmode` (k for kernel). Also, you can use the description field to point out the tristate, like: `-1=auto, 0=nokerncon, 1=kerncon`

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -999,6 +1037,35 @@ static int gdb_main_loop(struct gdb_excpt_ctx *g)
+	unsigned long irqs;
+
+	if (str) {
+		/* TODO: SMP Support */
```
I guess you save IRQs to be ready for preemption right?

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -999,6 +1037,35 @@ static int gdb_main_loop(struct gdb_excpt_ctx *g)
 	return (int)r;
 }
 
+__ssz gdb_virt_out(struct uk_console *dev __unused, const char *str, __sz len)
```
Can be `static`, right?

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -999,6 +1037,35 @@ static int gdb_main_loop(struct gdb_excpt_ctx *g)
+	return len;
+}
+
+__ssz gdb_virt_in(struct uk_console *dev __unused, char *buf __unused,
```
`static`?

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -1106,3 +1113,78 @@ int gdb_dbg_trap(int errnr, struct __regs *regs)
+{
+	/* Ignore breakpoints if the stub is not enabled on the command line. */
+	if (gdb_backing_console_id == GDB_BACKING_CONSOLE_NONE) {
+		uk_pr_info("Backing console not set\n");
```
I think we should not print in this case. If an image is compiled with GDB support but it is not enabled via command line, we should just silently continue.

skuenzer replied:
If you want to keep the message, you can use `uk_pr_debug`

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -1106,3 +1113,78 @@ int gdb_dbg_trap(int errnr, struct __regs *regs)
+	 * Maybe the backing console will be initialized later.
+	 */
+	if (gdb_backing_console_id >= uk_console_get_device_count()) {
+		uk_pr_info("Backing console not available\n");
```
`uk_pr_warn` or `uk_pr_err` instead? Maybe add also a bit more context (`gdbstub:`), like
`"gdbstub: Console device %d is not available\n"`

thass0 replied:
Doesn't it show in the debug output that the message is emitted from `gdbstub.c`? Why add the `gdbstub: `. Otherwise, I agree with adding more context

skuenzer replied:
True. Then just the message with `"con%"__PRIu16` instead of `Console device %d`

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -1106,3 +1113,78 @@ int gdb_dbg_trap(int errnr, struct __regs *regs)
+	/* Ignore breakpoints if the stub is not enabled on the command line. */
+	if (gdb_backing_console_id == GDB_BACKING_CONSOLE_NONE) {
+		uk_pr_info("Backing console not set\n");
+		return -ENXIO;
```
I would not consider this as error case.

thass0 replied:
I think it doesn't much matter here. Just look at the caller code of this function. I could also return a positive value instead of `-ENXIO` here. But we can't return `0` as there isn't actually a console that can be used.

skuenzer replied:
Hum... Maybe we can get it clearer if we move this check to `gdb_late_breakpoint()`, before calling this function. `gdb_backing_console_id == GDB_BACKING_CONSOLE_NONE` is saying that we do not want to have gdb stub waiting, correct?

thass0 replied:
Yes, that's correct

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -1106,3 +1113,78 @@ int gdb_dbg_trap(int errnr, struct __regs *regs)
+	return 0;
+}
+
+uk_plat_initcall_prio(gdb_late_breakpoint, 0, UK_PRIO_LATEST);
```
We could think of registering `gdb_late_breakpoint` to multiple inittab classes. This would enable early waiting n breakpoint for console drivers that are available earlier.

I think the breakpoint function should only once enter into a breakpoint.

thass0 replied:
Again, I would say that we don't want to go back to multiple stages of gdb stub initialization. This is what we just moved away from by removing the early init mechanism. `ukprint` will introduce a print buffer, right? Then output won't even get lost. My opposition to your comment here is similar to [this comment](https://github.com/unikraft/unikraft/pull/1479/commits/ffd403eb2cdb09c5ef0181a41214d6f307fa1fe3#r1758271414)

skuenzer replied:
As further discussed in https://github.com/unikraft/unikraft/pull/1479#discussion_r1758983716, nothing to change for now. All good :+1:

skuenzer commented on lib/ukdebug/include/uk/gdb.h:
```diff
@@ -0,0 +1,20 @@
+ * entered. If the setup is successful, this function will execute a breakpoint
+ * instruction to give the gdb host a chance to connect to the stub.
+ */
+void uk_debug_gdb_early_init(void);
```
As discussed offline, let us remove this for now since early init debugging can be a bit complicated due to order dependencies during early system initialization. This header is also not be needed anymore.

skuenzer commented on drivers/uktty/com/com.c:
```diff
@@ -194,12 +194,8 @@ __ssz com_out(struct uk_console *dev, const char *buf, __sz len)
 
 	com_dev = __containerof(dev, struct com_device, dev);
 
-	while (leftover--) {
```
This and the following changes regarding newline conversion. Somehow I am thinking they make more sense in the console PR #1464 . I let you decide if you want to move them over.

thass0 replied:
Let's just keep them here in this PR

skuenzer commented on lib/ukdebug/Config.uk:
```diff
@@ -136,4 +136,17 @@ config LIBUKDEBUG_GDBSTUB
 
+if LIBUKDEBUG_GDBSTUB
+config LIBUKDEBUG_GDB_ALWAYS_ACK
+	bool "Always use acknowledgment packets"
```
What do you think of: `Force "always ack" mode`

Should we also provide a kernel parameter option for this?

thass0 replied:
The only reason to touch this option is when using an unreliable serial connection. Idk if we need a kernel parameter to dynamically set this option. Your choice

thass0 replied:
I don't think "force" accurately describes what's going on. We don't force anything here. In fact, using acknowledgment packets is the default.

skuenzer replied:
I think the kernel parameter might be useful but you can keep it as is, as well (can be added later if needed). It would be interesting only for bare metal targets and we have currently a focus on virtualization environments.

It seems I did not well understand the GDB ACK option. Is the degree of ack'ing negotiated with gdb? Can also gdb ask for this and the option changes the initial default?

thass0 replied:
By default ack packets are always sent. The gbd host and the stub negotiate the available features right after connecting. At this stage, the stub can signal to the host that the stub wants to disable ack packets. Then, the gdb host can send the stub a packet to actually switch off ack packets. But this is up to the host; the stub can't control if this packet is sent by the stub or not. If ack packets are switched off, they are off forever. This is why even if we try to use "no ack" mode, it might turn out that the gdb host never sends us the packet to turn off ack packets, and as such ack packets stay on

skuenzer replied:
Thanks for the explanation. Makes much more sense now. As discussed offline we spoke about a slight update to the help summary to the option to make it slightly clearer.

thass0 committed:
lib/isrlib: Remove libc dependency

`isrlib` had a libc dependency because `isrlib` used libc-defined
types. However, there are equivalent types defined in the kernel
so that we can remove the libc dependency by switching to the
Unikraft-internal types. By removing the libc dependency, we also
remove the potential for a recursive dependency cycle. E.g., now
a library used by `nolibc` can safely select `isrlib`.

Checkpatch-Ignore: SPACING
Signed-off-by: Thassilo Schulze <thassilo@unikraft.io>

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -31,6 +53,15 @@ int gdb_dbg_getc(void)
 	return -ENOTSUP;
 }
 
+static int gdb_starts_with(const char *buf, __sz buf_len,
```
Can we find a slight better name for this function?

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -31,6 +53,15 @@ int gdb_dbg_getc(void)
 	return -ENOTSUP;
 }
 
+static int gdb_starts_with(const char *buf, __sz buf_len,
```
Can we find a slight better name for this function? We are doing here a matching of a prefix. Maybe: `gdb_memcmp_prefix`, `gdb_prefixcmp`, or `gdb_prefixmatch`?

thass0 replied:
`starts_with` is a common name for a function that matches a prefix. E.g., there's `starts_with` in Rust and `startswith` in Python

skuenzer replied:
👍🏻  Okay, fair enough

skuenzer commented on lib/ukdebug/arch/arm64/gdbsup.c:
```diff
@@ -10,9 +10,9 @@
-#include <uk/nofault.h>
 #include <uk/bitops.h>
 #include <uk/compiler.h>
+#include <uk/nofault.h>
```
Out of curiosity: Why this move in this commit?

thass0 replied:
No, reason. Probably happened by accident when rebasing for the nth time

thass0 replied:
I'm removing the move

skuenzer commented on lib/ukdebug/gdbstub.h:
```diff
@@ -0,0 +1,18 @@
+
+#ifndef __UKDEBUG_GDBSTUB_H__
+#define __UKDEBUG_GDBSTUB_H__
+
```
Always a good idea to include `uk/config.h` as well and first.

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -818,6 +818,76 @@ static int gdb_handle_q_cmd(char *buf, __sz buf_len,
+	char *buf_end = buf + buf_len;
+	int rc = 0;
+
+	/* We don't support the thread-related features of the gdb remote
```
Can this be tagged with: `NOTE:`?

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -998,6 +1038,34 @@ static int gdb_main_loop(struct gdb_excpt_ctx *g)
+
+	if (str) {
+		/* TODO: SMP Support */
+		irqs = ukplat_lcpu_save_irqf();
```
Why disabling interrupts and would it make more sense to move this to `gdb_send_message_packet()` and company?

thass0 replied:
As discussed offline, we keep this as a safety measure. See the comment in the code now.

skuenzer replied:
Thanks for the comment.

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -71,6 +72,33 @@ static __u16 gdb_backing_console_id = GDB_BACKING_CONSOLE_NONE;
+UK_LIBPARAM_PARAM_ALIAS(gdbmode, &gdb_virt_console_mode, int,
+			"-1=auto, 0=not shared 1=shared");
+
+static inline int gdb_virt_console_on(void)
```
Somehow, I find the naming of these configs checks not ideal when reading the function that uses those later. What would you think of:
`gdb_vcon_want_shared()` and `gdb_vcon_want_auto()`?

thass0 replied:
How about `gdb_mode_is_shared`/`gdb_mode_is_auto`?

skuenzer replied:
Good 👍🏻

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -1106,3 +1113,65 @@ int gdb_dbg_trap(int errnr, struct __regs *regs)
+static int gdb_late_breakpoint(struct uk_init_ctx *ctx __unused)
+{
+	/* This is the first time that the stub is entered. */
+	UK_ASSERT(!gdb_backing_console);
```
If we remove this assertion, we could call this init function multiple times right? I am thinking of patching it in a way that we try multiple times at various levels until we do not get `-ENXIO` anymore ... this way we would achieve kind of waiting for a console driver that is initializing late during init.

skuenzer replied:
It means that we need to skip successive calls when we connected once and keep retrying at each level if there was an issue before.

thass0 replied:
> I am thinking of patching it in a way that we try multiple times at various levels until we do not get -ENXIO anymore ... this way we would achieve kind of waiting for a console driver that is initializing late during init.

In principle, this is exactly what we did before when we had the early/late breakpoints split. We have purposefully decided against this split, so I don't see any reason to re-introduce such a mechanism. The `gdb_late_breakpoint` (it should be renamed `gdb_init_breakpoint` or something similar) function runs, all console drivers have been initialized.

skuenzer replied:
What I meant is slightly different: During early init we had more difficulties with solving dependencies. The inittab is different because when it is called, the most basic system is initialized and should allow us without much adoption to call gdb stub at every init stage. The idea was that an earlier call allows earlier debugging (including device drivers) - especially if we use the early console with this. Because we may also depend on a serial console device that is initialized later, I suggested the retry to find the earliest place. This currently happens latest in `UK_INIT_CLASS_PLAT`, via the `lib/ukbus` triggered device scanning. You register at the end of the same class, so it should be fine.
However, I am thinking now that there might be an even more elegant way than multiple registrations to each initlevel, so probably worth to move such changes to a different PR.

Agree with the renaming (to `gdb_entry`). 👍🏻  `gdb_stub_entry` may work as well but I assume we will add more ways to enter the debugger later. So, it is good for now as is.

michpappas replied:
To me that makes sense, as well as the proposal to look at it later. Perhaps in the scope of the topic of earlyboot, liblcpu and other items related to boot order. Either way, it's a good way to start gdb a bit earlier without dealing with the implications of early boot dependencies, especially given that we enable earlycon by default. Still, I think we don't need to register in every init class, just the very first one in case there is earlycon, and the regular one after consoles have been initialized, given that the priority of device initialization is deterministic.

skuenzer replied:
Agree.

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -1106,3 +1113,65 @@ int gdb_dbg_trap(int errnr, struct __regs *regs)
+	/* This is the first time that the stub is entered. */
+	UK_ASSERT(!gdb_backing_console);
+
+	uk_pr_info("Waiting for debugger ...\n");
```
Can we print the console device number as well, like:
`"Waiting for debugger connection at console %"__PRIu16"...\n"`

skuenzer commented on lib/ukdebug/gdbstub.c:
```diff
@@ -1106,3 +1113,65 @@ int gdb_dbg_trap(int errnr, struct __regs *regs)
+
+	uk_pr_info("Waiting for debugger ...\n");
+
+	if (gdb_setup_backing_console() < 0)
```
I would check the error code and be more verbose what happens, for instance:
`-ENXIO`: `uk_pr_warn("Console %"__PRIu16" does not exist or is not ready yet. Skipping waiting for debugger...\n")`

thass0 replied:
Checking the error code isn't really required, because this error message describes all possible error cases. I updated the error message though.

skuenzer replied:
Now it is this case, yes, I agree 👍🏻 .

michpappas reviewed (approved):
Great work and a much needed feature 🚀 

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

skuenzer reviewed (approved):
Thank you very much. Such a great feature and finally getting in.
Wanted since [2022](https://archive.fosdem.org/2022/schedule/event/skuenzer/) 😄 .

Reviewed-by: Simon Kuenzer <simon@unikraft.io>

skuenzer reviewed (approved):
Approved-by: Simon Kuenzer <simon@unikraft.io>
